### PR TITLE
Set ObservedGeneration in conditions

### LINKED ISSUE #1750: Set observedGeneration in Conditions
**What would you like to be added**:

Set the .observedGeneration in conditions based on .metadata.generation, in all Kueue objects.

**Why is this needed**:

From the k8s docs:

```
observedGeneration represents the .metadata.generation that the condition was set based upon.
For instance, if .metadata.generation is currently 12, but the .status.conditions[x].observedGeneration is 9,
the condition is out of date with respect to the current state of the instance.
```

It could be useful for understanding whether Kueue might still have to do other operations on the object

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by vladikkuzn**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Sets the .observedGeneration in conditions based on .metadata.generation, in all Kueue objects.

#### Which issue(s) this PR fixes:

Fixes #1750 

#### Does this PR introduce a user-facing change?

```release-note
Adds ObservedGeneration in conditions
```

### CODE REVIEW

File: pkg/workload/workload_test.go [hunk 1]
```diff
@@ -265,7 +263,12 @@ func TestUpdateWorkloadStatus(t *testing.T) {
 			if err := cl.Get(ctx, client.ObjectKeyFromObject(workload), &updatedWl); err != nil {
 				t.Fatalf("Failed obtaining updated object: %v", err)
 			}
-			if diff := cmp.Diff(tc.wantStatus, updatedWl.Status, ignoreConditionTimestamps); diff != "" {
+			if diff := cmp.Diff(
+				tc.wantStatus,
+				updatedWl.Status,
+				cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"),
+				cmpopts.IgnoreFields(metav1.Condition{}, "ObservedGeneration"),
```
Comment (trasc) [inline pkg/workload/workload_test.go:270]: Don't ignore the generation in unit tests since we should be able to control the value of the Generation field.

File: test/e2e/multikueue/e2e_test.go [hunk 2]
```diff
@@ -338,7 +339,7 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 						Reason:  "AllJobsCompleted",
 						Message: "jobset completed successfully",
 					},
-					cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"))))
+					util.IgnoreConditionTimestamps, util.IgnoreConditionObservedGeneration)))
```
Comment (trasc) [inline test/e2e/multikueue/e2e_test.go:342]: These are used together a lot, create something like `util.IgnoreConditionTimestampsAndObservedGeneration`

File: test/e2e/multikueue/e2e_test.go [hunk 3]
```diff
@@ -19,6 +19,8 @@ package mke2e
 import (
 	"os/exec"
 
+	"sigs.k8s.io/kueue/test/util"
+
```
Comment (trasc) [inline test/e2e/multikueue/e2e_test.go:23]: the order should be 

- standard
- external
- internal

move this import back down

File: test/e2e/singlecluster/e2e_test.go [hunk 4]
```diff
@@ -28,14 +28,15 @@ import (
 	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"sigs.k8s.io/kueue/test/util"
+
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
```
Comment (trasc) [inline test/e2e/singlecluster/e2e_test.go:33]: same

File: pkg/controller/admissionchecks/multikueue/admissioncheck.go [hunk 5]
```diff
@@ -70,10 +70,11 @@ func (a *ACReconciler) Reconcile(ctx context.Context, req reconcile.Request) (re
...
 	newCondition := metav1.Condition{
-		Type:    kueue.AdmissionCheckActive,
-		Status:  metav1.ConditionTrue,
-		Reason:  "Active",
-		Message: "The admission check is active",
+		Type:               kueue.AdmissionCheckActive,
+		Status:             metav1.ConditionTrue,
+		Reason:             "Active",
+		Message:            "The admission check is active",
+		ObservedGeneration: ac.Generation,
```
Comment (trasc) [inline pkg/controller/admissionchecks/multikueue/admissioncheck.go:77]: Make sure this change is covered in at least one unit test.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 6]
```diff
@@ -454,10 +454,11 @@ func (c *clustersReconciler) getKubeConfigFromPath(path string) ([]byte, bool, e
...
 	newCondition := metav1.Condition{
-		Type:    kueuealpha.MultiKueueClusterActive,
-		Status:  metav1.ConditionFalse,
-		Reason:  reason,
-		Message: message,
+		Type:               kueuealpha.MultiKueueClusterActive,
+		Status:             metav1.ConditionFalse,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: cluster.Generation,
```
Comment (trasc) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:461]: Make sure this change is covered in at least one unit test.

File: pkg/controller/admissionchecks/provisioning/admissioncheck_reconciler.go [hunk 7]
```diff
@@ -43,10 +43,11 @@ func (a *acReconciler) Reconcile(ctx context.Context, req reconcile.Request) (re
...
 	newCondition := metav1.Condition{
-		Type:    kueue.AdmissionCheckActive,
-		Status:  metav1.ConditionTrue,
-		Reason:  "Active",
-		Message: "The admission check is active",
+		Type:               kueue.AdmissionCheckActive,
+		Status:             metav1.ConditionTrue,
+		Reason:             "Active",
+		Message:            "The admission check is active",
+		ObservedGeneration: ac.Generation,
```
Comment (trasc) [inline pkg/controller/admissionchecks/provisioning/admissioncheck_reconciler.go:50]: Make sure this change is covered in at least one unit test.

File: pkg/controller/core/clusterqueue_controller.go [hunk 8]
```diff
@@ -650,10 +650,11 @@ func (r *ClusterQueueReconciler) updateCqStatusIfChanged(
...
 	meta.SetStatusCondition(&cq.Status.Conditions, metav1.Condition{
-		Type:    kueue.ClusterQueueActive,
-		Status:  conditionStatus,
-		Reason:  reason,
-		Message: msg,
+		Type:               kueue.ClusterQueueActive,
+		Status:             conditionStatus,
+		Reason:             reason,
+		Message:            msg,
+		ObservedGeneration: cq.Generation,
```
Comment (trasc) [inline pkg/controller/core/clusterqueue_controller.go:657]: Make sure this change is covered in at least one unit test.

File: pkg/controller/admissionchecks/multikueue/admissioncheck_test.go [hunk 9]
```diff
@@ -198,24 +255,36 @@ func TestReconcile(t *testing.T) {
...
+						Message:            `The admission check is active`,
+						ObservedGeneration: 1,
+					}).
+					// TODO: ask if this is correct
+					Condition(metav1.Condition{
+						Type:               kueue.AdmissionChecksSingleInstanceInClusterQueue,
+						Status:             metav1.ConditionTrue,
+						Reason:             SingleInstanceReason,
+						Message:            SingleInstanceMessage,
+						ObservedGeneration: 1,
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/admissioncheck_test.go:287]: It's better to just update `SingleInstanceInClusterQueue(true, SingleInstanceReason, SingleInstanceMessage).` to add the generation also.

File: pkg/scheduler/scheduler_test.go [hunk 10]
```diff
@@ -2232,7 +2230,7 @@ func TestRequeueAndUpdate(t *testing.T) {
 			if err := cl.Get(ctx, client.ObjectKeyFromObject(w1), &updatedWl); err != nil {
 				t.Fatalf("Failed obtaining updated object: %v", err)
 			}
-			if diff := cmp.Diff(tc.wantStatus, updatedWl.Status, ignoreConditionTimestamps); diff != "" {
+			if diff := cmp.Diff(tc.wantStatus, updatedWl.Status, cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime")); diff != "" {
```
Comment (trasc) [inline pkg/scheduler/scheduler_test.go:2233]: why this change?

File: pkg/controller/jobs/rayjob/rayjob_controller.go [hunk 11]
```diff
@@ -162,10 +162,11 @@ func (j *RayJob) RestorePodSetsInfo(podSetsInfo []podset.PodSetInfo) bool {
...
 	condition := metav1.Condition{
-		Type:    kueue.WorkloadFinished,
-		Status:  metav1.ConditionTrue,
-		Reason:  string(j.Status.JobStatus),
-		Message: j.Status.Message,
+		Type:               kueue.WorkloadFinished,
+		Status:             metav1.ConditionTrue,
+		Reason:             string(j.Status.JobStatus),
+		Message:            j.Status.Message,
+		ObservedGeneration: j.Generation,
```
Comment (trasc) [inline pkg/controller/jobs/rayjob/rayjob_controller.go:169]: Is this tested in a unit test?

File: pkg/controller/jobs/mpijob/mpijob_controller.go [hunk 12]
```diff
@@ -164,10 +164,11 @@ func (j *MPIJob) Finished() (metav1.Condition, bool) {
...
 	condition := metav1.Condition{
-		Type:    kueue.WorkloadFinished,
-		Status:  metav1.ConditionTrue,
-		Reason:  "JobFinished",
-		Message: message,
+		Type:               kueue.WorkloadFinished,
+		Status:             metav1.ConditionTrue,
+		Reason:             "JobFinished",
+		Message:            message,
+		ObservedGeneration: j.Generation,
```
Comment (trasc) [inline pkg/controller/jobs/mpijob/mpijob_controller.go:171]: Is this tested in a unit test?

File: pkg/controller/jobs/job/job_controller.go [hunk 13]
```diff
@@ -290,10 +290,11 @@ func (j *Job) Finished() (metav1.Condition, bool) {
...
 	condition := metav1.Condition{
-		Type:    kueue.WorkloadFinished,
-		Status:  metav1.ConditionTrue,
-		Reason:  "JobFinished",
-		Message: "Job finished successfully",
+		Type:               kueue.WorkloadFinished,
+		Status:             metav1.ConditionTrue,
+		Reason:             "JobFinished",
+		Message:            "Job finished successfully",
+		ObservedGeneration: j.Generation,
```
Comment (trasc) [inline pkg/controller/jobs/job/job_controller.go:297]: Is this tested in a unit test?

File: pkg/controller/jobframework/reconciler.go [hunk 14]
```diff
@@ -957,10 +957,11 @@ func generatePodsReadyCondition(job GenericJob, wl *kueue.Workload) metav1.Condi
...
 	return metav1.Condition{
-		Type:    kueue.WorkloadPodsReady,
-		Status:  conditionStatus,
-		Reason:  "PodsReady",
-		Message: message,
+		Type:               kueue.WorkloadPodsReady,
+		Status:             conditionStatus,
+		Reason:             "PodsReady",
+		Message:            message,
+		ObservedGeneration: wl.Generation,
```
Comment (trasc) [inline pkg/controller/jobframework/reconciler.go:964]: Is this tested in a unit test?

File: pkg/controller/core/localqueue_controller.go [hunk 15]
```diff
@@ -288,10 +288,11 @@ func (r *LocalQueueReconciler) UpdateStatusIfChanged(
...
 		meta.SetStatusCondition(&queue.Status.Conditions, metav1.Condition{
-			Type:    kueue.LocalQueueActive,
-			Status:  conditionStatus,
-			Reason:  reason,
-			Message: msg,
+			Type:               kueue.LocalQueueActive,
+			Status:             conditionStatus,
+			Reason:             reason,
+			Message:            msg,
+			ObservedGeneration: queue.Generation,
```
Comment (trasc) [inline pkg/controller/core/localqueue_controller.go:295]: Is this tested in a unit test?

File: pkg/workload/admissionchecks.go [hunk 16]
```diff
@@ -38,10 +38,11 @@ func SyncAdmittedCondition(w *kueue.Workload) bool {
...
 	newCondition := metav1.Condition{
-		Type:    kueue.WorkloadAdmitted,
-		Status:  metav1.ConditionTrue,
-		Reason:  "Admitted",
-		Message: "The workload is admitted",
+		Type:               kueue.WorkloadAdmitted,
+		Status:             metav1.ConditionTrue,
+		Reason:             "Admitted",
+		Message:            "The workload is admitted",
+		ObservedGeneration: w.Generation,
```
Comment (trasc) [inline pkg/workload/admissionchecks.go:45]: Is this tested in a unit test?

File: pkg/workload/workload.go [hunk 17]
```diff
@@ -408,6 +411,7 @@ func SetEvictedCondition(w *kueue.Workload, reason string, message string) {
 		LastTransitionTime: metav1.Now(),
 		Reason:             reason,
 		Message:            message,
+		ObservedGeneration: w.Generation,
```
Comment (trasc) [inline pkg/workload/workload.go:414]: Is this tested in a unit test?

File: pkg/workload/workload.go [hunk 18]
```diff
@@ -328,6 +328,7 @@ func UpdateStatus(ctx context.Context,
 		LastTransitionTime: now,
 		Reason:             reason,
 		Message:            api.TruncateConditionMessage(message),
+		ObservedGeneration: wl.Generation,
```
Comment (trasc) [inline pkg/workload/workload.go:331]: Is this tested in a unit test?

File: test/integration/controller/core/workload_controller_test.go [hunk 19]
```diff
@@ -269,10 +269,11 @@ var _ = ginkgo.Describe("Workload controller", ginkgo.Ordered, ginkgo.ContinueOn
...
 				}, util.Timeout, util.Interval).Should(gomega.BeComparableTo(&metav1.Condition{
-					Type:    kueue.WorkloadFinished,
-					Status:  metav1.ConditionTrue,
-					Reason:  "AdmissionChecksRejected",
-					Message: "Admission checks [check1] are rejected",
+					Type:               kueue.WorkloadFinished,
+					Status:             metav1.ConditionTrue,
+					Reason:             "AdmissionChecksRejected",
+					Message:            "Admission checks [check1] are rejected",
+					ObservedGeneration: createdWl.Generation,
```
Comment (trasc) [inline test/integration/controller/core/workload_controller_test.go:276]: Lets ignore the generation in the integration tests, and check the seating in the unit tests.

File: test/integration/controller/core/workload_controller_test.go [hunk 20]
```diff
@@ -314,10 +315,11 @@ var _ = ginkgo.Describe("Workload controller", ginkgo.Ordered, ginkgo.ContinueOn
...
 					g.Expect(createdWl.Status.Conditions).To(gomega.ContainElement(gomega.BeComparableTo(metav1.Condition{
-						Type:    kueue.WorkloadAdmitted,
-						Status:  metav1.ConditionTrue,
-						Reason:  "Admitted",
-						Message: "The workload is admitted",
+						Type:               kueue.WorkloadAdmitted,
+						Status:             metav1.ConditionTrue,
+						Reason:             "Admitted",
+						Message:            "The workload is admitted",
+						ObservedGeneration: createdWl.Generation,
```
Comment (trasc) [inline test/integration/controller/core/workload_controller_test.go:322]: Lets ignore the generation in the integration tests, and check the seating in the unit tests.

File: test/util/constants.go [hunk 21]
```diff
@@ -34,3 +37,9 @@ const (
 	ConsistentDuration = time.Second
 	Interval           = time.Millisecond * 250
 )
+
+var (
+	IgnoreConditionTimestamps                      = cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime")
```
Comment (trasc) [inline test/util/constants.go:42]: Is this used?

File: test/util/constants.go [hunk 22]
```diff
@@ -34,3 +37,9 @@ const (
...
+	IgnoreConditionTimestampsAndObservedGeneration = cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime", "ObservedGeneration")
+	IgnoreConditionMessage                         = cmpopts.IgnoreFields(metav1.Condition{}, "Message")
```
Comment (trasc) [inline test/util/constants.go:44]: Is this used?

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go [hunk 23]
```diff
@@ -145,25 +145,26 @@ func TestReconciler(t *testing.T) {
...
+											Name: "head-container",
+											Resources: corev1.ResourceRequirements{
+												Requests: make(corev1.ResourceList),
+											},
 										},
 									},
 								},
 							},
 						},
-					},
```
Comment (trasc) [inline pkg/controller/jobs/raycluster/raycluster_controller_test.go:166]: Is this doing anything?

**Review by trasc — COMMENTED** [review]
Just one nit.

Otherwise LGTM.

File: pkg/controller/jobs/mpijob/mpijob_controller_test.go
Comment (trasc) [inline pkg/controller/jobs/mpijob/mpijob_controller_test.go:1]: nit: I don't think this is needed, It looks to be testing that the added comment is not changing the resulting condition.

File: pkg/controller/jobs/rayjob/rayjob_controller_test.go
Comment (trasc) [inline pkg/controller/jobs/rayjob/rayjob_controller_test.go:1]: same as pkg/controller/jobs/mpijob/mpijob_controller_test.go

**Review by trasc — COMMENTED** [review]
LGTM overall. just two nits.

File: pkg/controller/jobs/mpijob/mpijob_controller_test.go [hunk 24]
```diff
@@ -322,3 +322,53 @@ func TestReconciler(t *testing.T) {
 	}
 
 }
+
+func TestFinished(t *testing.T) {
```
Comment (mimowo) [inline pkg/controller/jobs/mpijob/mpijob_controller_test.go]: What is the reason for adding the "negative" test cases? I would suggest to just drop them.

**Review by mimowo — COMMENTED** [review]
Lgtm overall, couple of nits

File: pkg/controller/jobs/raycluster/raycluster_controller_test.go [hunk 25]
```diff
@@ -145,25 +145,26 @@ func TestReconciler(t *testing.T) {
...
-						Count: int32(1),
-						Template: corev1.PodTemplateSpec{
-							Spec: corev1.PodSpec{
-								RestartPolicy: corev1.RestartPolicyNever,
-								Containers: []corev1.Container{
-									{
-										Name: "head-container",
-										Resources: corev1.ResourceRequirements{
-											Requests: make(corev1.ResourceList),
+					PodSets(
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/jobs/raycluster/raycluster_controller_test.go:150]: Ah, ok, there are not too many formatting changes here, so it may stay, but in general we try to do formatting in a dedicated PR, to minimize distractions.

File: pkg/controller/core/localqueue_controller_test.go [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/test/util"
+)
+
+func TestLocalQueueReconcile(t *testing.T) {
+	cases := map[string]struct {
+		clusterQueue   *kueue.ClusterQueue
+		localQueue     *kueue.LocalQueue
+		wantLocalQueue *kueue.LocalQueue
+		wantError      error
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/core/localqueue_controller_test.go:43]: Ok, got it

File: pkg/util/testingjobs/rayjob/wrappers.go [hunk 27]
```diff
@@ -159,3 +159,26 @@ func (j *JobWrapper) WorkloadPriorityClass(wpc string) *JobWrapper {
 	j.Labels[constants.WorkloadPriorityClassLabel] = wpc
 	return j
 }
+
```
Comment (mimowo) [inline pkg/util/testingjobs/rayjob/wrappers.go:162]: nit: do we still need these functions? Please drop unused.

**Comment by mimowo** [comment]
LGTM, last pending nit from me

**Comment by alculquicondor** [comment]
I don't remember if we added more conditions since you started this PR. Could you rebase and do a quick check?

**Comment by mimowo** [comment]
/lgtm
/approve
Thanks!
